onnxruntime/core/providers/coreml/builders/impl/cast_op_builder.cc

```diff
@@ -18,14 +19,47 @@ class CastOpBuilder : public BaseOpBuilder {
 
   bool HasSupportedInputsImpl(const Node& node, const OpBuilderInputParams& input_params,
                               const logging::Logger& logger) const override;
+
+ public:
+  bool SupportsMLProgram() const override { return true; }
 };
 
-Status CastOpBuilder::AddToModelBuilderImpl(ModelBuilder& /* model_builder */,
-                                            const Node& /* node */,
-                                            const logging::Logger& /* logger */) const {
-  // This is a special handling case for ArgMax Op, where argmax is followed by a cast to int32 type.
-  // The ArgMax is fused with the Cast node and produces an int32 output.
-  // Cast node is not provided in CoreML model, so we're skipping adding the Cast node here.
+Status CastOpBuilder::AddToModelBuilderImpl([[maybe_unused]] ModelBuilder& model_builder,
+                                            [[maybe_unused]] const Node& node,
+                                            [[maybe_unused]] const logging::Logger& logger) const {
+// This is a special handling case for ArgMax Op, where argmax is followed by a cast to int32 type.
+// The ArgMax is fused with the Cast node and produces an int32 output.
+#if defined(COREML_ENABLE_MLPROGRAM)
+  auto input_dtype = node.InputDefs()[0]->TypeAsProto()->tensor_type().elem_type();
+  // TensorProto_DataType_INT64 is not supported in CoreML MLProgram, only when the predeceased node is ArgMax
+  if (model_builder.CreateMLProgram() && input_dtype != ONNX_NAMESPACE::TensorProto_DataType_INT64) {
+    using namespace CoreML::Specification::MILSpec;
+    // https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS15.elementwise_unary.cast
+
+    std::unique_ptr<Operation> op = model_builder.CreateOperation(node, "cast");
+    AddOperationInput(*op, "x", node.InputDefs()[0]->Name());
+    NodeAttrHelper helper(node);
+    auto cast_to_type = helper.Get("to", ONNX_NAMESPACE::TensorProto::UNDEFINED);
+    std::string to_dtype = "";
+    if (cast_to_type == ONNX_NAMESPACE::TensorProto::INT32 || cast_to_type == ONNX_NAMESPACE::TensorProto::INT64) {
+      to_dtype = "int32";
+      cast_to_type = ONNX_NAMESPACE::TensorProto::INT32;
```

Comment:
nit: Might be good to add a comment here explaining that inside the CoreML model we can only use int32 as it doesn't support int64. Could also mention we convert CoreML model input/output between int32 and int64 when running it.